internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -900,6 +900,30 @@ public void broadcastTopicSwitch(
     producer.flush();
   }
 
+  /**
+   * Broadcast control message to real-time topic partition, to be consumed by venice leader. Partition high watermarks
+   * are left to local venice leader to prepare and then been produced to version topic partition.
+   *
+   * @param oldServingVersionTopic the version topic change capture consumer should switch from.
+   * @param newServingVersionTopic the version topic change capture consumer should switch to.
+   * @param debugInfo arbitrary key/value pairs of information that will be propagated alongside the control message.
+   */
+  public void broadcastVersionSwap(
```

Comment:
So this function is mainly for writing to RT since it doesn't carry any watermark info.

Regarding the leader to carry the watermark info to VT, I am wondering the actual usage.
Since the leader will process RT topics at different speed, and the timing of processing of `VS` CM from different RT will be different as well, are you thinking to produce one `VS` CM to VT whenever processing any `VS` CM from a RT? If not, during leadership handover, how will the leader know which `VS` CM from which RT has been processed? And if the leader produces a `VS` CM per RT, the new leader will be able to recover the full state during leadership handover, but that introduces one more problem since changelog consumer doesn't know which one is the latest `VS` CM when consuming VT, so should we introduce one more flag in `VS` CM to indicate whether it is the last `VS` CM or not?

Another item to consider is how the changelog consumer will handle this `VS` CM, and a native way is to do binary search of the new version topic since the `VS` CM doesn't carry the exact offset of the next VT topic.

Reply:
I think the initial goal is leader server will only produce `VS` CM to VT once, when they get all RT's `VS`. For basic support for one region's use case, there will be only one region's RT has data. So the leadership handover will not be an issue. 
Just want to make sure my understanding of your question is right. The leader handover you mentioned here is for new version push for the store of change capture enabled. 

To achieve multi-region consistency, these RT `VS`s are information to let old leader produce a `VS` CM to old change capture topic version's end (this `VS` CM will be used for change log consumer to switch to new change capture topic version) and then stop producing to the change capture topic. For each region, old leader will stop consuming each region's RT once it reads that region's `VS`. Old leader will keep tracking all regions' `VS` of all RTs.  
So until all region RTs' 'VS' are collected, otherwise the hand-over is not started.  

The new leader does not need which region's VS is read, it cares the HwM information shared globally from the old leader after it fully gathers all regions' RT `VS` and sharing HwM to indicate the handover is started, it starts produce to its corresponding change capure topic version. The changelog consumer can switch from previous's version's change capture topic based CM 'VS' message in the end of this topic.

Reply:
The leader handover I was referring to is the leadership handover of the current version topic in the middle of processing multiple RTs from different regions, and let us walk through some examples:
```
leader1: old leader
leader2: new leader
VS in RT1: at t1
VS in RT2: at t2.

At some point, leader1 consumes VS in RT1 and haven't consumed VS in RT2 yet, and now leadership handover happens, and leader2 will take it over from the last consumed upstream offsets, and the consumed offset will pass the VS in RT1 and before VS in RT2, so how would the new leader: leader2 know the VS in RT1 has already been processed? That is the reason I was thinking the leader should produce one VS message to VT whenever processing any VS message from any RT to carry over the state to the followers.
```
I hope it makes sense.

Regarding how version jump works in changelog consumer and it will be different for scenarios with or without before-image enabled.
Let us address the first issue before exploring the version jumping logic in changelog consumer.

Reply:
If the leader produces one `VS` CM to VT per `VS` message in RT and adding one more flag to indicate whether it is the last `VS` or not, the new leader will be able to recover the full state.
Regarding having all RT topics to be alive, this will become a different problem to solve. When this happens, it meant the version jump or the new future version will be delayed, and it might be an acceptable risk for the first iteration.

Reply:
I think this is related to which way we want to go, we have not decided yet, but seems `Version Swap Synchronization through real-time (RT) solution` has more issues than `controller based synchronization`. I agree that if we rely on RT to coordinate the synchronization, we need the field you mentioned.